crates/bevy_remote/src/lib.rs

```diff
@@ -0,0 +1,849 @@
+            data: None,
+        }
+    }
+}
+
+/// Error codes used by BRP.
+pub mod error_codes {
+    /// Invalid JSON.
+    pub const PARSE_ERROR: i16 = -32700;
+
+    /// JSON sent is not a valid request object.
+    pub const INVALID_REQUEST: i16 = -32600;
+
+    /// The method does not exist / is not available.
+    pub const METHOD_NOT_FOUND: i16 = -32601;
+
+    /// Invalid method parameter(s).
+    pub const INVALID_PARAMS: i16 = -32602;
+
+    /// Internal error.
+    pub const INTERNAL_ERROR: i16 = -32603;
+
+    // Bevy errors
+
+    /// Entity not found.
+    pub const ENTITY_NOT_FOUND: i16 = -23401;
+
+    /// Could not reflect or find component.
+    pub const COMPONENT_ERROR: i16 = -23402;
+
+    /// Could not find component in entity.
+    pub const COMPONENT_NOT_PRESENT: i16 = -23403;
+
+    /// Cannot reparent an entity to itself.
+    pub const SELF_REPARENT: i16 = -23404;
+}
+
+/// The result of a request.
+pub type BrpResult = Result<Value, BrpError>;
+
+/// The requests may occur on their own or in batches.
+/// Actual parsing is deferred for the sake of proper
+/// error reporting.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum BrpBatch {
+    /// Multiple requests with deferred parsing.
+    Batch(Vec<Value>),
+    /// A single request with deferred parsing.
+    Single(Value),
+}
+
+/// A message from the Bevy Remote Protocol server thread to the main world.
+///
+/// This is placed in the [`BrpMailbox`].
+#[derive(Debug, Clone)]
+pub struct BrpMessage {
+    /// The request method.
+    pub method: String,
+
+    /// The request params.
+    pub params: Option<Value>,
+
+    /// The channel on which the response is to be sent.
+    ///
+    /// The value sent here is serialized and sent back to the client.
+    pub sender: Sender<BrpResult>,
+}
+
+/// A resource that receives messages sent by Bevy Remote Protocol clients.
+///
+/// Every frame, the `process_remote_requests` system drains this mailbox and
+/// processes the messages within.
+#[derive(Debug, Resource, Deref, DerefMut)]
+pub struct BrpMailbox(Receiver<BrpMessage>);
+
+impl RemotePlugin {
+    /// Create a [`RemotePlugin`] with the default address and port but without
+    /// any associated methods.
+    fn empty() -> Self {
+        Self {
+            address: DEFAULT_ADDR,
+            port: DEFAULT_PORT,
+            methods: RwLock::new(vec![]),
+        }
+    }
+
+    /// Set the IP address that the server will use.
+    #[must_use]
+    pub fn with_address(mut self, address: impl Into<IpAddr>) -> Self {
+        self.address = address.into();
+        self
+    }
+
+    /// Set the remote port that the server will listen on.
+    #[must_use]
+    pub fn with_port(mut self, port: u16) -> Self {
+        self.port = port;
+        self
+    }
+
+    /// Add a remote method to the plugin using the given `name` and `handler`.
+    #[must_use]
+    pub fn with_method<M>(
+        mut self,
+        name: String,
```

Comment:
Nit: Any reason to not do `name: impl ToString` or `name: impl Into<String>`? I see `RemoteMethods::insert` at least uses the latter.

Reply:
Good idea!